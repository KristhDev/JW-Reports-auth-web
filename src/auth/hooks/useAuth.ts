/* Supabase */
import { supabase } from '../../supabase';

/**
 * Hooks to manage the user authentication.
 */
const useAuth = () => {
    /**
     * This function is responsible for updating the user's password, for this it refreshes the
     * session that was sent with the token, with the session already active it updates the
     * userpassword and then logs out.
     * @param {string} newPassword - The new password you want to set
     * @param {string} token - The refresh token that was sent to the user's email.
     * @returns The error message.
     */
    const updatePassword = async (newPassword: string, token: string) => {
        const { error } = await supabase.auth.refreshSession({ refresh_token: token });

        if (error) {
            console.error(error);
            return error.message;
        }

        const { error: passError } = await supabase.auth.updateUser({ password: newPassword });

        if (passError) {
            console.error(passError);
            return passError.message;
        }

        const { error: logoutError } = await supabase.auth.signOut();

        if (logoutError) {
            console.error(logoutError);
            return logoutError.message;
        }

        window.location.reload();
    }

    return {
        updatePassword
    }
}

export default useAuth;