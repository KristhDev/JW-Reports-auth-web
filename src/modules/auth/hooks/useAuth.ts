/* Supabase */
import { supabase } from '../../../config';

/**
 * Hooks to manage the user authentication.
 */
const useAuth = () => {

    /**
     * This function is responsible for updating the user's password. It refreshes the session that was sent with the token,
     * updates the user's password, and then logs out.
     *
     * @param {string} newPassword - The new password you want to set.
     * @param {string} token - The refresh token that was sent to the user's email.
     * @return {Promise<string | void>} The error message if there was an error, otherwise undefined.
     */
    const updatePassword = async (newPassword: string, token: string): Promise<string | void> => {
        const { error } = await supabase.auth.refreshSession({ refresh_token: token });

        if (error) {
            console.error(error);
            return error.message;
        }

        const { error: passError } = await supabase.auth.updateUser({ password: newPassword });

        if (passError) {
            console.error(passError);
            return passError.message;
        }

        const { error: logoutError } = await supabase.auth.signOut();

        if (logoutError) {
            console.error(logoutError);
            return logoutError.message;
        }

        window.location.reload();
    }

    return {
        updatePassword
    }
}

export default useAuth;