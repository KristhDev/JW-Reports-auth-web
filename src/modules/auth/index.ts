export * from './components';
export * from './hooks';
export * from './layouts';
export * from './pages';
export * from './routes';