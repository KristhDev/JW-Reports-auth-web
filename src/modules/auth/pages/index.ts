export { Confirm } from './Confirm';
export { ResetPassword } from './ResetPassword';