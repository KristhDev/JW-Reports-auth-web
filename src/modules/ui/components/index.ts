export { Button } from './Button';
export { EyeBtn } from './EyeBtn';
export { FormField } from './FormField';
export { Spinner } from './Spinner';
export { Title } from './Title';