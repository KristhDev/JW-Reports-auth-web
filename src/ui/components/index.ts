export { FormField } from './FormField';
export { Title } from './Title';