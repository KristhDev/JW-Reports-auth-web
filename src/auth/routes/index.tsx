import { RouteObject } from 'react-router-dom';

import { Confirm, ResetPassword } from '../pages';

export const routes: RouteObject[] = [
    {
        path: '',
        element: <Confirm />
    },
    {
        path: 'reset-password',
        element: <ResetPassword />
    }
];