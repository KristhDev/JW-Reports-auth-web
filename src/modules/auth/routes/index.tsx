import { RouteObject } from 'react-router-dom';

/* Pages */
import { Confirm, ResetPassword } from '../pages';

/* Creating a route object. */
export const routes: RouteObject[] = [
    {
        path: '',
        element: <Confirm />
    },
    {
        path: 'reset-password',
        element: <ResetPassword />
    }
];