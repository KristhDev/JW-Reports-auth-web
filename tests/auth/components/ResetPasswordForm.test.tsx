import { Mock } from 'vitest';
import { render, screen, fireEvent, act, waitFor } from '@testing-library/react';
import { RouterProvider, createMemoryRouter } from 'react-router-dom';

import { ResetPasswordForm } from '../../../src/auth/components';

import { useAuth } from '../../../src/auth/hooks';

vi.mock('../../../src/auth/hooks/useAuth.ts');

const updatePasswordMock = vi.fn();

const passwordRouter = createMemoryRouter([
    {
        path: '/reset-password',
        element: <ResetPasswordForm />
    }
], { initialEntries: [ '/reset-password#refresh_token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiI' ] });

describe('Test in <ResetPasswordForm /> component', () => {
    (useAuth as Mock).mockReturnValue({
        updatePassword: updatePasswordMock
    });

    beforeEach(() => {
        vi.clearAllMocks();
    });

    it('should to match snapshot', () => {
        const { container } = render(<RouterProvider router={ passwordRouter } />);
        expect(container).toMatchSnapshot();
    });

    it('should not call updatePassword then form is empty', async () => {
        render(<RouterProvider router={ passwordRouter } />);

        const form = await screen.findByTestId<HTMLFormElement>('form');
        fireEvent.submit(form);

        expect(updatePasswordMock).not.toHaveBeenCalled();
    });

    it('should not call updatePassword then is invalid', async () => {
        await waitFor(() => {
            render(<RouterProvider router={ passwordRouter } />);
        });

        const inputPassword = await screen.findByPlaceholderText<HTMLInputElement>('Ingrese su nueva contraseña');
        const inputConfirmPassword = await screen.findByPlaceholderText<HTMLInputElement>('Confirme su nueva contraseña');

        act(() => {
            fireEvent.input(inputPassword, { target: { value: 'test' } });
            fireEvent.input(inputConfirmPassword, { target: { value: 'test-error' } });
        });

        const form = await screen.findByTestId<HTMLFormElement>('form');
        fireEvent.submit(form);

        expect(updatePasswordMock).not.toHaveBeenCalled();
    });

    it('should call updatePassword', async () => {
        await waitFor(() => {
            render(<RouterProvider router={ passwordRouter } />);
        });

        const inputPassword = await screen.findByPlaceholderText<HTMLInputElement>('Ingrese su nueva contraseña');
        const inputConfirmPassword = await screen.findByPlaceholderText<HTMLInputElement>('Confirme su nueva contraseña');

        act(() => {
            fireEvent.input(inputPassword, { target: { value: 'test-error' } });
            fireEvent.input(inputConfirmPassword, { target: { value: 'test-error' } });
        });

        const form = await screen.findByTestId<HTMLFormElement>('form');
        fireEvent.submit(form);

        await waitFor(() => {
            expect(updatePasswordMock).toHaveBeenCalledTimes(1);
            expect(updatePasswordMock).toHaveBeenCalledWith('test-error', 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiI');
        });
    });

    it('should render error message of submit', async() => {
        updatePasswordMock.mockReturnValue('Error message');

        await waitFor(() => {
            render(<RouterProvider router={ passwordRouter } />);
        });

        const inputPassword = await screen.findByPlaceholderText<HTMLInputElement>('Ingrese su nueva contraseña');
        const inputConfirmPassword = await screen.findByPlaceholderText<HTMLInputElement>('Confirme su nueva contraseña');

        act(() => {
            fireEvent.input(inputPassword, { target: { value: 'test-error' } });
            fireEvent.input(inputConfirmPassword, { target: { value: 'test-error' } });
        });

        const form = await screen.findByTestId<HTMLFormElement>('form');
        fireEvent.submit(form);

        const submitError = await screen.findByText('Error message');
        expect(submitError).toBeTruthy();
    });
});