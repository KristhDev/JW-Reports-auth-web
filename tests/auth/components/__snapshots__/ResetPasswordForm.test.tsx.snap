// Vitest Snapshot v1

exports[`Test in <ResetPasswordForm /> component > should to match snapshot 1`] = `
<div>
  <div
    class="form"
  >
    <form
      action="#"
      class="form__container"
      data-testid="form"
    >
      <div
        class="form-field__container"
      >
        <label
          class="form-field__label undefined"
          data-testid="form-field-label"
          for="password"
        >
          Nueva contrase単a:
        </label>
        <div
          class="form-field__input-container"
        >
          <input
            class="form-field__input-container__input undefined"
            name="password"
            placeholder="Ingrese su nueva contrase単a"
            type="password"
            value=""
          />
          <button
            class="eye-btn"
          >
            <svg
              class="eye-btn__icon"
              data-testid="eye-icon"
              fill="currentColor"
              height="25"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="25"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M255.66 112c-77.94 0-157.89 45.11-220.83 135.33a16 16 0 00-.27 17.77C82.92 340.8 161.8 400 255.66 400c92.84 0 173.34-59.38 221.79-135.25a16.14 16.14 0 000-17.47C428.89 172.28 347.8 112 255.66 112z"
                fill="none"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="32"
              />
              <circle
                cx="256"
                cy="256"
                fill="none"
                r="80"
                stroke-miterlimit="10"
                stroke-width="32"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="form-field__container"
      >
        <label
          class="form-field__label undefined"
          data-testid="form-field-label"
          for="confirmPassword"
        >
          Confirmar contrase単a:
        </label>
        <div
          class="form-field__input-container"
        >
          <input
            class="form-field__input-container__input undefined"
            name="confirmPassword"
            placeholder="Confirme su nueva contrase単a"
            type="password"
            value=""
          />
          <button
            class="eye-btn"
          >
            <svg
              class="eye-btn__icon"
              data-testid="eye-icon"
              fill="currentColor"
              height="25"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="25"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M255.66 112c-77.94 0-157.89 45.11-220.83 135.33a16 16 0 00-.27 17.77C82.92 340.8 161.8 400 255.66 400c92.84 0 173.34-59.38 221.79-135.25a16.14 16.14 0 000-17.47C428.89 172.28 347.8 112 255.66 112z"
                fill="none"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="32"
              />
              <circle
                cx="256"
                cy="256"
                fill="none"
                r="80"
                stroke-miterlimit="10"
                stroke-width="32"
              />
            </svg>
          </button>
        </div>
      </div>
      <button
        class="button undefined"
        style="margin-top: 2rem;"
        type="submit"
      >
        Restablecer
      </button>
    </form>
  </div>
</div>
`;
